### DISCUSSION TITLE
Separate platform logic in the core project

### DISCUSSION THREAD

**Discussion by LSXAxeller** [AvaloniaUI/Avalonia#14916] [Q&A] [answered]
I have a project separated into four projects using the Avalonia MVVM template with .NET 8:
- NovelNodePlayer
- NovelNodePlayer.Android
- NovelNodePlayer.Browser
- NovelNodePlayer.Desktop

In NovelNodePlayer, all logic is stored. However, in NovelNodePlayer.Android, the screen resolution isn't as good as on Desktop, despite being the same. Additionally, NovelNodePlayer.Browser freezes the task on await Task.Delay, while NovelNodePlayer.Desktop works fine.

The main problem is that I have code in NovelNodePlayer where I want to execute one part if the platform is Android, and another part if it's not. I've created the provided script, but it doesn't work on Android. by the way, Is there a way to debug from an Android phone, such as notifying the debug terminal with a message or error?

```
using System.IO;
using System.Threading.Tasks;
using NAudio.Wave;
#if ANDROID // __ANDROID__ or ANDROID both doesn't working
using Android.Media;
#endif

namespace NovelNodePlayer.Data
{
    public class AudioPlayer : IDisposable
    {
        public AssetPlayState PlaybackState { get; private set; } = AssetPlayState.Stopped;

        private readonly bool loop;
        private bool isDisposed;

        private readonly IWavePlayer waveOut;
        private readonly MemoryStream audioStream;
        private readonly string audioPath;

        public AudioPlayer(byte[] audioData, string path = ".mp3", bool loop = false)
        {
            if (audioData == null || audioData.Length == 0)
            {
                isDisposed = true;
                return;
            }

            this.loop = loop;
            PlaybackState = AssetPlayState.Stopped;
            audioPath = path;

            audioStream = new MemoryStream(audioData);

#if ANDROID // __ANDROID__
            waveOut = new AudioTrackOut();
#else
            waveOut = new WaveOutEvent();
#endif

        }

        public void Play()
        {
            if (isDisposed) return;
        
            if (PlaybackState != AssetPlayState.Playing)
            {
                waveOut.Play();
                PlaybackState = AssetPlayState.Playing;
            }
        }
    }
}

#if ANDROID // __ANDROID__
// Android wave player implemention by @neilt6, https://gist.github.com/neilt6/6d07322070470536ea0ba409c343c2a5
namespace NAudio.Wave
{
    /// <summary>
    /// Represents an Android wave player implemented using <see cref="AudioTrack"/>.
    /// </summary>
    public class AudioTrackOut : IWavePlayer
    {
		// Code removed for shortening
    }
}
#endif
```

I am calling from another script in NovelNodePlayer project too
```
                var audioPlayer = new AudioPlayer(aduioData);
                audioPlayer.Play();
```

**Comment by maxkatz6** [comment, marked as answer]
You have two approaches:
1. Extract implementation from interfaces. And put platform specific implementation into platform specific projects.
2. Make shared project multitargetted, i.e. replace `<TargetFramework>net8.0</TargetFramework>` with `<TargetFrameworks>net8.0-android;net8.0</TargetFrameworks>`. And then you can use `#if ANDROID` directives in the same project.

**Comment by thevortexcloud** [comment]
> NovelNodePlayer.Browser freezes the task on await Task.Delay,

https://github.com/AvaloniaUI/Avalonia/issues/14022

> Is there a way to debug from an Android phone

You should be able to launch the app in debug mode if you have it plugged into your PC via a cable. You can then set breakpoints in your IDE. It wont be as good as debugging a desktop app, but it's better than nothing. Otherwise you can use standard android logcat stuff.

Anyway. your approach will work but is fairly nightmarish to maintain. Avoid using preprocessors whenever possible. It can quickly spiral out of control due to the compiler skipping all code on a different preprocessor branch. I would suggest abstracting things and using DI instead.

https://docs.avaloniaui.net/docs/guides/implementation-guides/how-to-implement-dependency-injection
